### PR TITLE
KAFKA-14509: [4/4] Handle includeAuthorizedOperations

### PR DESCRIPTION
**Pull request by riedelmax**
Last PR for KAFKA-14509

### CODE REVIEW

**Comment by dajac** [comment]
@riedelmax Thanks for the patch. Could you please extend unit and integration tests to cover this change?

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 1]
```diff
@@ -3852,6 +3853,17 @@ class KafkaApis(val requestChannel: RequestChannel,
         if (exception != null) {
           requestHelper.sendMaybeThrottle(request, consumerGroupDescribeRequest.getErrorResponse(exception))
         } else {
+          if (request.header.apiVersion >= 3 && includeAuthorizedOperations) {
```
Comment (dajac) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: I suppose that we don't need the version check because this API only has version zero.

**Comment by chia7712** [comment]
out of curiosity, do we have IT for that option?  I grep code base and it seems the related ITs are running with old coordinator/protocol. For example:

1. https://github.com/apache/kafka/blob/fb566e48bf05d749f8db8da803a3570acf25bb11/core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala#L1213
2. https://github.com/apache/kafka/blob/fb566e48bf05d749f8db8da803a3570acf25bb11/core/src/test/scala/integration/kafka/api/DescribeAuthorizedOperationsTest.scala#L139

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 2]
```diff
@@ -7086,6 +7086,7 @@ class KafkaApisTest extends Logging {
   def testConsumerGroupDescribe(): Unit = {
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: Should we parameterize the tests to test both variants?

**Review by dajac — COMMENTED** [review]
@riedelmax Thanks for the update. I left one more comment. I also wonder if we could add a test to `ConsumerGroupDescribeRequestsTest` to cover the change. Have you considered it?

**Comment by riedelmax** [comment]
@dajac thanks for checking so quickly. I will have time on friday to do the integration tet

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala (hunk 2 above)
Comment (riedelmax) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: good idea

**Comment by riedelmax** [comment]
@chia7712 im not sure if I understand you correctlu. We have some integration tests in `ConsumerGroupDescribeRequestTest`

**Comment by chia7712** [comment]
> We have some integration tests in ConsumerGroupDescribeRequestTest

thanks for this reminder. I neglect that before :(

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 3]
```diff
@@ -7107,11 +7108,12 @@ class KafkaApisTest extends Logging {
     ).asJava)
 
     // Can't reuse the above list here because we would not test the implementation in KafkaApis then
+    val authorizedOperationsInt = if (includeAuthorizedOperations) 328 else Int.MinValue;
     val describedGroups = List(
       new DescribedGroup().setGroupId(groupIds.get(0)),
       new DescribedGroup().setGroupId(groupIds.get(1)),
       new DescribedGroup().setGroupId(groupIds.get(2))
-    ).map(group => group.setAuthorizedOperations(328)) // Integer representation of authorized operations for this request
+    ).map(group => group.setAuthorizedOperations(authorizedOperationsInt)) // Integer representation of authorized operations for this request
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: nit: Should we move this comment next to `328`?

**Comment by dajac** [comment]
@riedelmax There are a bunch of conflicts. Could you please fix them? Regarding `ConsumerGroupDescribeRequestsTest`, is it possible to extend it too?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala (hunk 2 above)
Comment (riedelmax) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: @dajac Do you think it is necessary to have two tests, one with includeAuthorizedOperations and one without it? 
I could at it to the IT, without having one IT with includeAuthorizedOperations=false.
Its covered in the unit tests anyways and the option does not change the behavior of the API

Comment (dajac) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: Yeah, adding it to the existing integration test is fine for me.

File: core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestsTest.scala [hunk 4]
```diff
@@ -116,6 +115,7 @@ class ConsumerGroupDescribeRequestsTest(cluster: ClusterInstance) extends GroupC
     val timeoutMs = 5 * 60 * 1000
     val clientId = "client-id"
     val clientHost = "/127.0.0.1"
+    val authorizedOperationsInt = 328; // Integer representation of the authorized operations for this request
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestsTest.scala]: I wonder if we should replace `328` by something like
```
      Utils.to32BitField(
        AclEntry.supportedOperations(ResourceType.CLUSTER).asScala
          .map(_.code.asInstanceOf[JByte]).asJava)
```
What do you think? We could also do it in the other place.

Comment (riedelmax) [reply core/src/test/scala/unit/kafka/server/ConsumerGroupDescribeRequestsTest.scala]: done

**Comment by chia7712** [comment]
@riedelmax please rebase code to have the fix #16249

**Comment by riedelmax** [comment]
@chia7712 would you mind to explain how #16249 effects this PR? if there are no conflicts with trunk i dont need to rebase or merge manually right?

**Comment by chia7712** [comment]
@riedelmax #16249 fix the blocked tests. Without that fix, the CI will get timeout when running your PR

**Comment by chia7712** [comment]
I prefer to check all failed tests before merging. And the last commit of this PR does not have completed CI. That is why I suggest to rebase code to trigger QA again.

**Comment by dajac** [comment]
Merged trunk to include https://github.com/apache/kafka/pull/16249.

**Review by dajac — APPROVED** [review]
LGTM, thanks. We need to cherry-pick it to 3.8 too.

**Comment by dajac** [comment]
CI does not seem to be working...

**Comment by dajac** [comment]
Merged to trunk and to 3.8.

**Pull request merged**